### PR TITLE
Upcast conjure elemental

### PR DESCRIPTION
**Pull request by PhilPJL**
Stage 1 done and seems to work ok.  Maybe don't merge unless Chris is happy with this approach.
Stage 2 is to add CR6 and 7 versions of the fire, air, earth elementals - need some stats

### CODE REVIEW

**Review by DubhHerder — APPROVED** [review]
Looks fine to me, especially if it's a quick implementation as noted. One thing that might cause confusion for some users is calling some of the elementals huge when they are still large size creature but that's really minor and doesn't effect implementation. I didn't look at the scaling in terms of balance etc

File: SolastaCommunityExpansion/Models/UpcastSummonsContext.cs [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+            {
+                return new MonsterBuilder(name, CreateGuid(name), title, description, baseMonster);
+            }
+
+            string CreateGuid(string name) => GuidHelper.Create(Namespace, name).ToString("N");
+        }
+
+        internal static void ApplyUpcastSummon(EffectForm effectForm, int effectiveLevel)
+        {
+            if (!Main.Settings.EnableUpcastConjureElemental
```
Comment (ChrisPJohn) [inline SolastaCommunityExpansion/Models/UpcastSummonsContext.cs:228]: What's the plan for scaling this to support other summon spells (like fey as the comment states above)? I feel like either the setting should be more general (like EnableUpcastSummoningSpells) or there should be a check for for if the spell is conjure elemental.

File: SolastaCommunityExpansion/Models/UpcastSummonsContext.cs [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+
+                if (effectiveLevel <= upcastSummonInfo.OriginalSpellLevel)
+                {
+                    Main.Log($"UpcastSummon-ApplySpellLevel: {effectiveLevel} <= {upcastSummonInfo.OriginalSpellLevel} - ignoring");
+                    return;
+                }
+
+                if (effectiveLevel - upcastSummonInfo.OriginalSpellLevel > upcastSummonInfo.UpcastMonsterDefinitionNames.Length)
+                {
+                    Main.Log($"UpcastSummon-ApplySpellLevel: {effectiveLevel}, no suitable upcast - ignoring");
```
Comment (ChrisPJohn) [inline SolastaCommunityExpansion/Models/UpcastSummonsContext.cs:253]: Should this use the highest available? So if the spell is 3rd level and there's upcasts for 4 and 5 but not 6, if I use a 6th level slot I should get the 5th level version.

**Review by ChrisPJohn — CHANGES_REQUESTED** [review]

File: SolastaCommunityExpansion/Models/UpcastSummonsContext.cs (hunk 1 above)
Comment (PhilPJL) [reply SolastaCommunityExpansion/Models/UpcastSummonsContext.cs:228]: I change this to just Elementals for now and then rethink when I see what TA finalize for Conjure Fey

File: SolastaCommunityExpansion/Models/UpcastSummonsContext.cs (hunk 2 above)
Comment (PhilPJL) [reply SolastaCommunityExpansion/Models/UpcastSummonsContext.cs:253]: Good point.

Comment (PhilPJL) [reply SolastaCommunityExpansion/Models/UpcastSummonsContext.cs]: Done

**Comment by PhilPJL** [comment]
Works as is, and I've bumped up the elemental to hit and damage bonus.

File: SolastaCommunityExpansion/Builders/MonsterAttackDefinitionBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        }
+
+        public MonsterAttackDefinitionBuilder SetToHitBonus(int value)
+        {
+            Definition.SetToHitBonus(value);
+            return this;
+        }
+
+        public MonsterAttackDefinitionBuilder SetDamageBonus(int value)
+        {
```
Comment (DubhHerder) [inline SolastaCommunityExpansion/Builders/MonsterAttackDefinitionBuilder.cs:20]: This seems a bit targeted and specific for a builder, if it's included in the builders, maybe the name could mention that it only changes the first effect form damage bonus. It wouldn't work on monster attacks effect descriptions that use any other effect form  in the first entry or effect descriptions with multiple damage entries.

**Review by DubhHerder — COMMENTED** [review]
Just a small note on the monster attack builders

File: SolastaCommunityExpansion/Builders/MonsterAttackDefinitionBuilder.cs (hunk 3 above)
Comment (PhilPJL) [reply SolastaCommunityExpansion/Builders/MonsterAttackDefinitionBuilder.cs:20]: I realised that :)  Need something like

```
        public MonsterAttackDefinitionBuilder SetDamageBonus(int value)
        {
            var effectForm = Definition.EffectDescription.GetFirstFormOfType(EffectForm.EffectFormType.Damage);
            effectForm?.DamageForm.SetBonusDamage(value);

            return this;
        }

```

Comment (PhilPJL) [reply SolastaCommunityExpansion/Builders/MonsterAttackDefinitionBuilder.cs:20]: I meant to look at what you'd done but forgot.

Comment (DubhHerder) [reply SolastaCommunityExpansion/Builders/MonsterAttackDefinitionBuilder.cs:20]: I did pretty much the same as yourself (copy an existing attack and alter it's parameters) just without the builder as i often wanted to change more than just the first damage bonus entry by adding new effect forms, changing the secondary damage bonuses, die types, damage types etc. 
If i was doing it again, i'd consider using Chris's approach of new effect descriptions etc via the general builders and effectdescription.copy(). 
For reference, here's where i defined monster attacks  but probably not much there you don't already know
https://github.com/ThyWoof/SolastaDungeonMakerPro/blob/master/SolastaDungeonMakerPro/Models/MonsterHelpers/NewMonsterAttacks.cs

Comment (PhilPJL) [reply SolastaCommunityExpansion/Builders/MonsterAttackDefinitionBuilder.cs:20]: Ha, I don't know much, I've left all the new monsters, classes etc to Chris and you.

**Review by ChrisPJohn — APPROVED** [review]

**Comment by PhilPJL** [comment]
Going for subspell filter.

**Pull request closed without merging**